### PR TITLE
[pipeline-to-taskrun] Add custom task that runs a Pipeline as a TaskRun 🍙

### PR DESCRIPTION
**Pull request by bobcatfish**
# Changes

This commit adds a custom task that allows user to run simple sequential
Pipelines as one TaskRun - which means the Pipeline can refer to
multiple Tasks but run on only one pod.

It only supports a subset of Pipeline functionality (more detail on what
and why in the README) but is enough that folks can do a lot of what
they would have previously used PipelineResources for, e.g. doing a git
clone and then doing something with the data, in the same pod, and emit
results such as the exact commit sha used.

Next steps will be to expand the functionality supported, get feedback,
and if the feedback is good, promote this to a top level Pipeline API
feature.

Experimental project proposal: https://github.com/tektoncd/community/issues/447

Added @wlynch @jerop  as owners also, just ping me if you've changed your minds XD

p.s. shout out to the yaml test lib @imjasonh  - i had to do some truly terrible things to make it work sometimes but it was still SOOOO much better than creating go structs XD

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [x] Commit messages includes a project tag in the subject - e.g. [OCI], [hub], [results], [task-loops]

_See [the contribution guide](https://github.com/tektoncd/experimental/blob/master/CONTRIBUTING.md)
for more details._

### CODE REVIEW

**Review by jerop — COMMENTED** [review]
thanks for the detailed documentation!! 

please add `pipeline-to-taskrun` here to trigger its tests: https://github.com/tektoncd/experimental/blob/6ae091569832eca4d96c493e03897f146074e534/test/presubmit-tests.sh#L61

can't see its tests in the [logs](https://storage.googleapis.com/tekton-prow/pr-logs/pull/tektoncd_experimental/770/pull-tekton-experimental-integration-tests/1413648890258788352/build-log.txt)

**Comment by bobcatfish** [comment]
Thanks for catching that @jerop!!

**Comment by bobcatfish** [comment]
/me slowly watches all the tests fail

File: pipeline-to-taskrun/pkg/reconciler/pipelinetotaskrun/pipelinetotaskrun.go [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+		return fmt.Errorf("couldn't create taskrun for pipeline %s: %v", run.Spec.Ref.Name, err)
+	}
+
+	logger.Infof("Creating a new TaskRun object %s", tr.Name)
+	if _, err := r.pipelineClientSet.TektonV1beta1().TaskRuns(run.Namespace).Create(ctx, tr, metav1.CreateOptions{}); err != nil {
+		logger.Errorf("Run %s/%s got an error creating TaskRun - %v", run.Namespace, run.Name, err)
+		run.Status.MarkRunFailed(ReasonRunFailedCreatingPipelineRun,
+			"Run got an error creating pipelineRun - %v", err)
+		return err
+	}
```
Comment (jerop) [inline pipeline-to-taskrun/pkg/reconciler/pipelinetotaskrun/pipelinetotaskrun.go:178]: following this section is a bit confusing i.e. it's not clear when taskrun is created 

e.g. we have an error "couldn't create taskrun for pipeline", then later we're "creating a new taskrun object"

may help if the first part used `merge` and second part part used `create`

Comment (jerop) [reply pipeline-to-taskrun/pkg/reconciler/pipelinetotaskrun/pipelinetotaskrun.go:178]: or you could add comments like in the other sections above, they were really helpful - thanks!

**Review by jerop — COMMENTED** [review]
looks great! so excited to see this feature 😀

File: pipeline-to-taskrun/pkg/reconciler/pipelinetotaskrun/pipelinetotaskrun.go (hunk 1 above)
Comment (bobcatfish) [reply pipeline-to-taskrun/pkg/reconciler/pipelinetotaskrun/pipelinetotaskrun.go:178]: both great suggestions! ive made a couple updates but lemme know if it's not enough of a change and i could still make it more clear

**Comment by bobcatfish** [comment]
Should be ready for another look @jerop ! 🙏

**Review by jerop — APPROVED** [review]
very exciting! thank you @bobcatfish 😁 

/lgtm

**Comment by jerop** [comment]
@afrittoli looks like I'm not allowed to approve because the owners file is not yet in, the top level owners have to approve 

cc @tektoncd/experimental-maintainers

**Review by jerop — APPROVED** [review]
thanks @bobcatfish 😁 

/lgtm

**Pull request merged**